synapse/storage/databases/main/search.py

```diff
@@ -760,20 +781,204 @@ def _to_postgres_options(options_dict: JsonDict) -> str:
+                    tokens.append(SearchToken.And)
+
+                # Add the search term.
+                tokens.append(word)
+
+    return tokens
+
+
+def _tokens_to_tsquery(tokens: TokenList) -> str:
+    """
+    Convert the list of tokens to a string suitable for passing to postgresql's to_tsquery
+
+    Ref: https://www.postgresql.org/docs/current/textsearch-controls.html
+    """
+
+    tsquery = []
+    for i, token in enumerate(tokens):
+        if isinstance(token, str):
+            tsquery.append(token)
+        elif isinstance(token, Phrase):
+            tsquery.append(
+                " ( "
+                + "".join(
+                    t if isinstance(t, str) else f" <{t.proximity}> "
+                    for t in token.phrase
+                )
+                + " ) "
+            )
+        elif token == SearchToken.Not:
+            tsquery.append(" !")
+        elif token == SearchToken.Or:
+            tsquery.append(" | ")
+        elif token == SearchToken.And:
+            tsquery.append(" & ")
+        else:
+            raise ValueError(f"unknown token {token}")
+
+        if (
+            i != len(tokens) - 1
+            and isinstance(token, (str, Phrase))
+            and tokens[i + 1] not in (SearchToken.Or, SearchToken.And)
+        ):
+            tsquery.append(" & ")
+    return "".join(tsquery)
+
+
+def _tokens_to_sqlite_match_query(tokens: TokenList) -> str:
+    """
+    Convert the list of tokens to a string suitable for passing to sqlite's MATCH.
+    Assume sqlite was compiled with enhanced query syntax.
+
+    Ref: https://www.sqlite.org/fts3.html#full_text_index_queries
+    """
+    match_query = []
+    for token in tokens:
+        if isinstance(token, str):
+            match_query.append(token)
+        elif isinstance(token, Phrase):
+            # SQLite doesn't support followed by operators. NEAR is close, but
+            # discards order.
+            match_query.append(
+                '"' + " ".join(t for t in token.phrase if isinstance(t, str)) + '"'
+            )
+        elif token == SearchToken.Not:
+            # TODO: SQLite treats NOT as a *binary* operator. Hopefully a search
+            # term has already been added before this.
+            match_query.append(" NOT ")
+        elif token == SearchToken.Or:
+            match_query.append(" OR ")
+        elif token == SearchToken.And:
+            match_query.append(" AND ")
+        else:
+            raise ValueError(f"unknown token {token}")
+
+    return "".join(match_query)
+
+
+def _parse_query_for_pgsql(search_term: str, engine: PostgresEngine) -> Tuple[str, str]:
+    """Selects a tsquery_* func to use and transforms the search_term into syntax appropriate for it.
+
+    Args:
+        search_term: A user supplied search query.
+        engine: The database engine.
+
+    Returns:
+        A tuple of (parsed search_term, tsquery func to use).
+    """
+
+    if engine.supports_websearch_to_tsquery:
```

Comment:
How awful would it be to not do all the fancy handling of and vs or etc for old postgres? Especially given PG10 is EoL in a fortnight...

Reply:
I think we could remove a bunch of code. We could just fallback to `plain_to_tsquery` or something on old postgres, which won't work great but shouldn't break anything.